### PR TITLE
[DBatch] dynamic batching graph rewrite

### PR DESCRIPTION
**Pull request by szha**

### CODE REVIEW

File: src/imperative/batching.cc [hunk 1]
```diff
@@ -59,28 +63,132 @@ nnvm::Graph DBatchEngine::BatchGraphs(std::vector<nnvm::Graph>& graphs) {
...
 
   // generate new graph
-  // (TODO szha)
+  // create mapping from old node to new node
+  std::unordered_map<nnvm::Node*, std::vector<nnvm::NodePtr>> old_new_node_map;
+  for (uint32_t istep = 0; istep < forward_steps.size(); istep++) {
+    const std::unordered_map<int64_t, std::vector<nnvm::NodePtr>> step = forward_steps[istep];
+    for (const std::pair<int64_t, std::vector<nnvm::NodePtr>> step_ops : step) {
+      int64_t op_sign = step_ops.first;
+      const std::vector<nnvm::NodePtr> op_ptrs = step_ops.second;
```
Comment (ZihengJiang) [inline src/imperative/batching.cc]: const T&

File: include/mxnet/batching.h [hunk 2]
```diff
@@ -84,9 +84,12 @@ class DBatchEngine {
   nnvm::NodeEntryMap<NDArray> entry_arr_;
 
 
-  nnvm::Graph BatchGraphs(std::vector<nnvm::Graph>& graphs);
-  void ExecuteGraph(nnvm::Graph& graph);
+  nnvm::Graph BatchGraphs(const std::vector<nnvm::Graph>& graphs);
+  void ExecuteGraph(const nnvm::Graph& graph);
 };
+nnvm::NodeEntry map_node_entry(const nnvm::NodeEntry entry,
```
Comment (ZihengJiang) [inline include/mxnet/batching.h]: do not expose in header file

File: src/imperative/batching.cc [hunk 3]
```diff
@@ -59,28 +63,132 @@ nnvm::Graph DBatchEngine::BatchGraphs(std::vector<nnvm::Graph>& graphs) {
...
+            concat_node_inputs.emplace_back(map_node_entry(input_entry,
+                                                           old_new_node_map));
+            concat_node_name += "_" + input_entry.node->attrs.name;
+          }
+          std::unordered_map<std::string, std::string> concat_args = {
+            {"num_args", std::to_string(num_nodes)},
+            {"dim", std::to_string(in_batch_axis)}
+          };
+
+          batch_node_inputs.emplace_back(MakeNode("concat", "concat_", concat_node_inputs,
```
Comment (ZihengJiang) [inline src/imperative/batching.cc]: maybe move CreateConcatNode outside as a function

Comment (ZihengJiang) [reply src/imperative/batching.cc]: also add cache for batched input

File: src/imperative/batching.cc [hunk 4]
```diff
@@ -43,12 +43,29 @@ int64_t ComputeNodeSign(const nnvm::NodePtr& n) {
...
+}
+
+nnvm::Graph DBatchEngine::BatchGraphs(const std::vector<nnvm::Graph>& graphs) {
   // collect depth
   int max_depth = 0;
   std::unordered_map<nnvm::Node*, int> depth_map;
+  std::vector<nnvm::NodeEntry> prev_outputs;
   for (const nnvm::Graph& g : graphs) {
-    nnvm::DFSVisit(g.outputs, [&](const nnvm::NodePtr& n){
+    std::vector<nnvm::NodeEntry> g_outputs = g.outputs;
```
Comment (eric-haibin-lin) [inline src/imperative/batching.cc]: const ref

File: src/imperative/batching.cc [hunk 5]
```diff
@@ -43,12 +43,29 @@ int64_t ComputeNodeSign(const nnvm::NodePtr& n) {
   return reinterpret_cast<int64_t>(n->op());
 }
 
-nnvm::Graph DBatchEngine::BatchGraphs(std::vector<nnvm::Graph>& graphs) {
+nnvm::NodeEntry map_node_entry(const nnvm::NodeEntry entry,
+                               std::unordered_map<nnvm::Node*,
```
Comment (eric-haibin-lin) [inline src/imperative/batching.cc]: ref

File: src/imperative/batching.cc [hunk 6]
```diff
@@ -43,12 +43,29 @@ int64_t ComputeNodeSign(const nnvm::NodePtr& n) {
...
+                                                  std::vector<nnvm::NodePtr>> node_map) {
+  nnvm::Node* node_ptr = entry.node.get();
```
Comment (eric-haibin-lin) [inline src/imperative/batching.cc]: Const

File: src/imperative/batching.cc [hunk 7]
```diff
@@ -43,12 +43,29 @@ int64_t ComputeNodeSign(const nnvm::NodePtr& n) {
...
+  if (node_map.find(node_ptr) == node_map.end()) {
```
Comment (eric-haibin-lin) [inline src/imperative/batching.cc]: Iter

File: src/imperative/batching.cc [hunk 8]
```diff
@@ -43,12 +43,29 @@ int64_t ComputeNodeSign(const nnvm::NodePtr& n) {
...
+    return entry;
+  }
+  std::vector<nnvm::NodePtr> new_nodes = node_map[node_ptr];
+  nnvm::NodePtr result_node;
+  result_node = new_nodes[entry.index < new_nodes.size() ? entry.index : 0]; // possibly non-batchable multi-out node
```
Comment (eric-haibin-lin) [inline src/imperative/batching.cc]: Comment

File: src/imperative/batching.cc [hunk 9]
```diff
@@ -59,28 +76,119 @@ nnvm::Graph DBatchEngine::BatchGraphs(std::vector<nnvm::Graph>& graphs) {
...
+  std::unordered_map<nnvm::Node*, std::vector<nnvm::NodePtr>> old_new_node_map;
+  for (uint32_t istep = 0; istep < forward_steps.size(); istep++) {
+    const std::unordered_map<int64_t, std::vector<nnvm::NodePtr>> step = forward_steps[istep];
+    for (const std::pair<int64_t, std::vector<nnvm::NodePtr>> step_ops : step) {
+      int64_t op_sign = step_ops.first;
+      const std::vector<nnvm::NodePtr>& op_ptrs = step_ops.second;
+      size_t num_nodes = op_ptrs.size();
+      nnvm::Node* first_op_node = op_ptrs.front().get();
+      if (num_nodes == 1) { // op that can't be batched
+        std::vector<nnvm::NodeEntry> old_inputs = first_op_node->inputs;
```
Comment (eric-haibin-lin) [inline src/imperative/batching.cc]: Ref

File: src/imperative/batching.cc [hunk 10]
```diff
@@ -59,28 +76,119 @@ nnvm::Graph DBatchEngine::BatchGraphs(std::vector<nnvm::Graph>& graphs) {
...
+      }
+
+      size_t num_inputs = first_op_node->num_inputs(), num_outputs = first_op_node->num_outputs();
+
+      std::unordered_set<size_t> batchable_data_indices = {0}; // TODO find batchable data tensors from op
+
+      // record lengths of each sample
+      std::vector<size_t> sample_lengths(num_nodes);
+      for (uint32_t j = 0; j < num_nodes; j++) {
+        sample_lengths[j] = 1; // TODO get shape from node->ndarray mapping
```
Comment (eric-haibin-lin) [inline src/imperative/batching.cc]: Finish todo

File: src/imperative/batching.cc (hunk 3 above)
Comment (szha) [reply src/imperative/batching.cc]: cache look-up can be expensive. will not fix.

**Review by ZihengJiang — APPROVED** [review]

**Pull request merged**